### PR TITLE
[flang] Change `uniqueCGIdent` separator from `.` to `X`

### PR DESCRIPTION
**Pull request by fabianmcg**
Change the separator in the `uniqueCGIdent` method to `X`. This change is required to enable OpenMP offloading for the NVPTX target, as dots are not valid identifiers in PTX and `uniqueCGIdent` is used to mangle some literals. A follow up patch will add support for the NVPTX target.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/FIRBuilder.cpp [hunk 1]
```diff
@@ -966,13 +966,13 @@ std::string fir::factory::uniqueCGIdent(llvm::StringRef prefix,
     llvm::SmallString<32> str;
     llvm::MD5::stringifyResult(result, str);
     std::string hashName = prefix.str();
-    hashName.append(".").append(str.c_str());
+    hashName.append("_").append(str.c_str());
     return fir::NameUniquer::doGenerated(hashName);
   }
   // "Short" identifiers use a reversible hex string
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Builder/FIRBuilder.cpp:972]: Is there some change required in the demangling code?

File: flang/lib/Optimizer/Builder/FIRBuilder.cpp [hunk 2]
```diff
@@ -966,13 +966,13 @@ std::string fir::factory::uniqueCGIdent(llvm::StringRef prefix,
...
   std::string nm = prefix.str();
   return fir::NameUniquer::doGenerated(
-      nm.append(".").append(llvm::toHex(name)));
+      nm.append("_").append(llvm::toHex(name)));
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Builder/FIRBuilder.cpp:975]: The document might have to be updated.
https://github.com/llvm/llvm-project/blob/main/flang/docs/BijectiveInternalNameUniquing.md

**Review by kiranchandramohan — COMMENTED** [review]
Thanks for the patch. Please wait for comments from others before proceeding.

**Review by jeanPerier — COMMENTED** [review]
Thanks, this looks good to me.

Thanks @kiranchandramohan for pointing out the documentation and demangling. The constant litteral name are "QQ" internal name that are not meant to be demangled to some symbol + scope, and we do not document the mangling of the internal QQ name (they do not really matter from a Fortran ABI point of view since constant literals should be local to compilation units). The documentation only mentions QQ is reserved for internal names, so it is still valid after this change.

@fabianmcg, note that I see at least a couple of other cases where dots may be emitted in assembly names.
- When dealing with array literals (as opposed to scalars like in this patch): https://github.com/llvm/llvm-project/blob/58679ea576f9e48e20e43dcd99fd75c98fb7e6ba/flang/lib/Lower/Mangler.cpp#L283
- runtime type info is currently using dots (only matters if some runtime using derived type descriptors make it to the device I guess). https://github.com/llvm/llvm-project/blob/58679ea576f9e48e20e43dcd99fd75c98fb7e6ba/flang/test/Lower/derived-type-descriptor.f90#L14

**Comment by fabianmcg** [comment]
> note that I see at least a couple of other cases where dots may be emitted in assembly names.

@jeanPerier  I was planning to split `.` -> `_` changes into multiple patches, should I bundle all mangling changes in this patch?

**Comment by jeanPerier** [comment]
If it makes things easier, I think a more generic solution could also be to deal with this when dealing with the target aspects of the IR in the stage that lowers FIR to LLVM IR like the ExternalNameConversion pass https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/Transforms/ExternalNameConversion.cpp.

FIR symbol names do not have to be the assembly names already (and are not in the case of external subroutine/common blocks).

Although I am OK with the character and array litteral change in FIR that you have, they do not change the complexity and readability. Dealing with the derived type descriptor data may be a bit more complex to avoid collision with user defined symbols. And yes, it is OK with me and better to split the patches since they are touching a lot of test files.

**Comment by vzakhari** [comment]
@fabianmcg, can you please clarify if it is a requirement that in the final host and device "images" the variables are named the same way? Will there be any `RegisterVar` call in the host code that takes the variable's device name?

**Comment by fabianmcg** [comment]
@jeanPerier one option is adding a pass exclusive to NVPTX that renames symbols with dots. I just thought that having a uniform naming scheme was better, I'm open to other options.

> @fabianmcg, can you please clarify if it is a requirement that in the final host and device "images" the variables are named the same way? Will there be any `RegisterVar` call in the host code that takes the variable's device name?

I'm not 100% sure. Strictly speaking, in LLVM IR there are no hard requirements and one only needs to refer to the correct name, so they can be named differently. However, I don't know if there are additional limitations imposed by the current implementation of the OMP dialect and the OpenMP IRbuilder.

One thing to mention is that this issue is not being caused by mapping of symbols, but instead by the addition of literals to the device module, with these literal having invalid identifiers.

**Comment by fabianmcg** [comment]
@jdoerfert just pointed out that `$` is a better option for offload identifiers as there is no possibilities of collision with user defined names -literals are never going to collide.  I see 2 options:
1. Change all separators to `$` or some other valid separator for the sake of consistency.
2. Adding a pass to change `.` in offload identifiers.

**Comment by jdoerfert** [comment]
> @jdoerfert just pointed out that `$` is a better option for offload identifiers as there is no possibilities of collision with user defined names -literals are never going to collide. I see 2 options:
> 
> 1. Change all separators to `$` or some other valid separator for the sake of consistency.
> 2. Adding a pass to change `.` in offload identifiers.

Since QQ is reserved, we can use `_` with QQ, assuming all use cases have QQ or some other way of preventing clashes.

Wrt. renaming late, I generally advise against. It makes tracing names across stages unnecessarily complicated. Consistent names (throughout the pipeline and across targets) makes tooling and debugging much easier.

**Comment by jeanPerier** [comment]
> Since QQ is reserved, we can use `_` with QQ, assuming all use cases have QQ or some other way of preventing clashes.

The second case I mentioned above ([derived type runtime description data](https://github.com/llvm/llvm-project/blob/58679ea576f9e48e20e43dcd99fd75c98fb7e6ba/flang/test/Lower/derived-type-descriptor.f90#L14)) is read only data, but is not mangled with QQ currently because it is created by semantics as compiler created symbol in user scopes and is transparent in lowering. The '.' currently ensures they cannot collide with user symbols.

Regarding '$', are we sure that all assembly languages are OK with it? For instance, googling for some ISA, the IBM AIX Version 7.2 defines that "Symbols may consist of numeric digits, underscores, periods, uppercase or lowercase letters, or any
combination of these." (page 28 of https://www.ibm.com/docs/en/ssw_aix_72/assembler/assembler_pdf.pdf). I am not an expert, so I am not sure this matters, but before changing '.' to '$', it should be ensured it is valid on every llvm targets, or this should be target specific mangling/renaming.

**Comment by fabianmcg** [comment]
> Regarding '$', are we sure that all assembly languages are OK with it?

@jeanPerier that's a good point we'd end up circling back if we encounter  `$` is not valid in a given target. What about an upper case separator, or a series of upper case chars?

1. `.` -> `X` ?
2. `.` -> `X_X` ?

I'm suggesting `X` because I think it's not in use.

**Comment by jeanPerier** [comment]
> @jeanPerier that's a good point we'd end up circling back if we encounter `$` is not valid in a given target. What about an upper case separator, or a series of upper case chars?
> I'm suggesting `X` because I think it's not in use.

I think it is a good idea. Adding @vdonaldson here to double check that there will not be collisions with this strategy.

**Comment by vdonaldson** [comment]
Given the discussion to this point, I'm inclined to favor `X`. That should avoid collisions.

The derived type info use case may be the most significant, the most common, and the most visible. There are likely a few other places that use `.` in compiler-generated symbol names beyond those already mentioned. There are two instances of `.list` suffixes in `lib/Lower/IO.cpp`.  The first one is [here](https://github.com/llvm/llvm-project/blob/58679ea576f9e48e20e43dcd99fd75c98fb7e6ba/flang/lib/Lower/IO.cpp#L324).

**Comment by fabianmcg** [comment]
Switched `.` to `X`, follow up patches will change the remainder of `.` appearances to `X` and updating the docs.

**Review by jeanPerier — APPROVED** [review]
Thanks for working on the issue and for the update. I think you may have forgotten to push the updated mangling doc to mention X is now used as a separator in internal names. Otherwise looks great.

Also, you may want to locally rebase and test your PR with check-flang before merging to ensure that no new tests with QQcl. made it in the main branch during the review.

**Pull request merged**